### PR TITLE
Rework game server health initial delay handling

### LINKED ISSUE #2966: Agones health check shouldn't fail during game server container image pull
**What happened**:

The gameserver-sidecar health check fails while the game server container image is being pulled.

**What you expected to happen**:

The health check should not fail at this time, allowing image pull to complete without terminating the game server.

**How to reproduce it (as minimally and precisely as possible)**:

Given the following example config, the health check will fail 105 seconds after start if the game server isn't healthy.

```
health:
    failureThreshold: 3
    initialDelaySeconds: 60
    periodSeconds: 15
```

Unfortunately, this doesn't account for the amount of time that it takes for the container image to be pulled, which can exceed 105 seconds in some cases.  For example, if an image pull takes 3 minutes, the health check will fail and attempt to terminate the pod after 105 seconds. 

**Anything else we need to know?**:

This behavior traces back to Agones v1.19, when https://github.com/googleforgames/agones/issues/2355 was merged which altered the order of container startup so that Agones starts first and the game server starts second.  This causes the delay timer to start earlier than previously.

A workaround is to increase the `initialDelaySeconds` to a larger value, the longest we expect image pull to take.  This prevents the health check from failing and terminating the game server, but configuring the delay that large introduces a blind spot in monitoring.  If image pulls take less than the delay, such as in the case where the image already exists locally, then there is no health monitoring during the remaining duration until the `initialDelaySeconds` expires.

**Environment**:
- Agones version: v1.28
- Kubernetes version (use `kubectl version`): v1.23
- Cloud provider or hardware configuration: AWS
- Install method (yaml/helm): Helm
- Troubleshooting guide log(s):
- Others:

### LINKED ISSUE #3058: CI flaky on Autopilot 1.26
Seeing lots of builds where the e2es are failing on the Autopilot 1.26 cluster. Debugging now.

### PR DESCRIPTION
**Pull request by zmerlynn**
Implements https://github.com/googleforgames/agones/issues/2966#issuecomment-1492346086:

* We remove any knowledge in the SDK of InitialDelaySeconds
* We remove the runHealth goroutine from main and shift this responsibility to the /gshealthz handler

Along the way:

*  I noted that the FailureThreshold doesn't need to be enforced on both the kubelet and SDK side, so in the injected liveness probe, I dropped that to 1. Previously we were waiting more probes than we needed to. In practice this is not terribly relevant since the SDK pushes it into Unhealthy.

Does not try to fully enforce the SDK start order as noted in the same comment (which may be racy). I researched that more and can't find a way I like - using a command means we have to do something gross to pick between unix/windows, since the SDK is supported for Windows pods as well. Still thinking on this part.

Fixes #2966 
Fixes #3058

### CODE REVIEW

**Comment by zmerlynn** [comment]
Go is over-zealously detecting a data race here. I'll fix it shortly, but not tonight.

File: pkg/sdkserver/sdkserver.go [hunk 1]
```diff
@@ -267,12 +238,69 @@ func (s *SDKServer) Run(ctx context.Context) error {
...
+
+	mux.HandleFunc("/healthz", func(w http.ResponseWriter, r *http.Request) {
+		_, err := w.Write([]byte("ok"))
+		if err != nil {
+			s.logger.WithError(err).Error("could not send ok response on healthz")
+			w.WriteHeader(http.StatusInternalServerError)
+		}
+	})
+	mux.HandleFunc("/gshealthz", func(w http.ResponseWriter, r *http.Request) {
+		s.gameServerContainerStarted(initialDelay)
```
[the remark this answers is not in the document]
Comment (zmerlynn) [reply pkg/sdkserver/sdkserver.go:260]: I think I replied to that in https://github.com/googleforgames/agones/issues/2966#issuecomment-1478300425, but I am now realizing I missed part of your point.

The issue is that right now the SDK starts healthchecking immediately, running this function periodically: https://github.com/googleforgames/agones/blob/main/pkg/sdkserver/sdkserver.go#L820-L826

We could do something simple here and change it so that the `s.healthy()` clause is only even checked if we've seen `/gshealthz`, e.g. `/gshealthz` setting a flag that the container has started. You're right that that might be simpler.

[the remark this answers is not in the document]
Comment (markmandel) [reply pkg/sdkserver/sdkserver.go:260]: It might be possible to use a `sync.Once()` to handle the initial starting operation as well.

But yes, that's exactly my thought 👍🏻

[the remark this answers is not in the document]
Comment (zmerlynn) [reply pkg/sdkserver/sdkserver.go:260]: > It might be possible to use a sync.Once() to handle the initial starting operation as well.

Yeah, I'm toying with this for sure.

[the remark this answers is not in the document]
Comment (zmerlynn) [reply pkg/sdkserver/sdkserver.go]: PTAL!

File: pkg/sdkserver/sdkserver.go [hunk 2]
```diff
@@ -335,7 +327,7 @@ func (s *SDKServer) updateState(ctx context.Context) error {
 
 	// If we are setting the Allocated status, set the last-allocated annotation as well.
 	if gsCopy.Status.State == agonesv1.GameServerStateAllocated {
-		ts, err := s.clock.Now().MarshalText()
+		ts, err := s.clock.Now().UTC().MarshalText()
```
Comment (markmandel) [inline pkg/sdkserver/sdkserver.go:330]: if we're going to do this here, we also need to do it in Allocate() functionality as well, so it's UTC everywhere.

**Review by markmandel — COMMENTED** [review]
The approach looks good - but looking at the e2e tests, it does seem like there's a regression somewhere.

Have you tried running the e2e tests individually on your local machine?

File: pkg/sdkserver/sdkserver.go (hunk 2 above)
Comment (zmerlynn) [reply pkg/sdkserver/sdkserver.go:330]: I think all of the failures can be attributed to `TestPlayerConnectWithCapacityZero`, which I have #3062 out for. The remaining failures are unreported status because that test exploded. I'm seeing failures in that test in other builds as well, so I'm not certain it's this PR, and I did try to repro with that test locally, but let me try again.

**Comment by zmerlynn** [comment]
> The approach looks good - but looking at the e2e tests, it does seem like there's a regression somewhere.
> 
> Have you tried running the e2e tests individually on your local machine?

(Meant to reply here.)

I think all of the failures can be attributed to `TestPlayerConnectWithCapacityZero`, which I have #3062 out for. The remaining failures are unreported status because that test explodes (and takes out the remaining parallel tests). I'm seeing failures in that test in other builds as well, so I'm not certain it's this PR, and I did try to repro with that test locally, but let me try again.

File: pkg/sdkserver/sdkserver.go (hunk 2 above)
Comment (zmerlynn) [reply pkg/sdkserver/sdkserver.go:330]: Let me revert this part.

**Comment by zmerlynn** [comment]
> but let me try again.

20 tries in and `TestPlayerConnectWithCapacityZero` refuses to fail locally, though I'm running it in isolation. Digging in a little on the full e2e case.

**Comment by zmerlynn** [comment]
You're right there may be some real issue, though. From [raw e2e logs of a failing case](https://ff13601f75b78463dfa12eca2ca613298c2837c690cfcf1ee5de67d-apidata.googleusercontent.com/download/storage/v1/b/258182270954.cloudbuild-logs.googleusercontent.com/o/log-301d82e6-2b78-4955-a1ce-fd0250f4b5b4.txt?jk=Ac_6HjKAvrMGlhRAj_6B0PZCuyRyVn_aoerAPQ_q_8YAkyrwUnJW50tL6fhqGSPgyp4bzCqhNMAo0usG8EKVlHaCcRCMjSE8V5WmODJUpn77lF1Rn_FjbRYTMoB5fxk-gaAZntt9ri-vVoUIZDRrXeroZYEDYbnGaoKd9k5pxrmyY_iSgWmPFiIWbZ0ueBIXdPqF8bgQ3j2rfgFN_5RFiKHYG5WqleTw5oV1GamtXZ6XisY9C-gtVF7x7-zSuNfvx3gHcLz7jUSE9FKtxR3Q6uWdIJkUquph7t3Rtb5scEHiyXYIpX1b8sd91egp6wOvxV-jrAXdeDQhDnXuUiNBthO5RA4PWRayGNmWbdlKyrcCNgWvm1axI3zt_mPL9d9eTAyCb-UX3InCXlxoljc72Zgl4TFyU8fg49f_qY08y-S-M_PKt2ZfcblvOXABkHBs7-XGB1lPpIiRLFlpuUVzJKjFJSXkLL20US1Nj9qWpeJB3UClaHKef3LI8Gt0zVFrG06O45PxgTdX2qGodeid576hwfX2SyfOvgD6BghyWnjhAVpPYsbHn2dC4wwfzSV_Xn8pN58Lt2Admm4-5BDl0ncG8AYslAuQs4Rhf9XMWAWCdJCi1DWkzqSbilOWcQlZwY8IRUiM7W5-XFnKgyeVgVgLjjOjoxmK-T_o-J8FmB3PAM_LJMif_3XBQlW3w1myQ_Vs3v9s84fflODomlghLBzdXGPqKZw5DxwMebU8-C0Rvcp3_Fg8-stLEfQ8SQDCqF56SqwmaRYfdmzy29ykXU__xXYWNsfGbsm0Cz88ZKPmN7ka0kYaIQYEWLuHCTqKBmQ1OIgbe_lN2jqlQjdkhEfL178E-AZO2xphTXdchmLNYNHyat_whNFb6knLwYrVom3IbPW6yOf0fZ5_1LpW9ax3vxWKHO4PctdnuK3P2RU1Q-QQXfJZX_0aA6IjwTbXidoPOujhg1siSgz7WyVcMx8LmeD2bywPWxSpHi3wOE_PHyXxAEewYuMIoQFh5EyhozKtA2JhDGPz9CH1yWWTcSxgmtBD4tjTeCJMXlRTdjzWULClBumc0-SxGivr1coylQQEcpKdE6zqyhHH1C8VYwG2aZElFGknZoomRGBab7R_gE0AL6lXfu3mnPWG1peIzwoFuN9qWNbYG5V5YECcdwm0s8wudcd10bkvOJUBkJHC4g3hB2UEzC28LiIs0ljMyayUnYjSKdOXWiOUeRhNXQr-x9mjvwdnZZquDI_1FLozB6Zh8mA&isca=1):

```
Step #1 - "e2e-feature-gates": time="2023-03-31 22:07:53.321" level=info msg="Failed to send UDP packet to GameServer. Dumping Events!" gs=game-servers4l44 status="{State:Ready Ports:[{Name:udp-port Port:7128}] Address:34.83.136.12 NodeName:gke-gke-standard-e2e-test-clu-default-653e1e73-k0ft ReservedUntil:<nil> Players:0xc00087d050 Counters:map[] Lists:map[] Eviction:<nil>}" test=TestPlayerConnectWithCapacityZero
Step #1 - "e2e-feature-gates": time="2023-03-31 22:07:53.321" level=info msg="Dumping Events:" kind= test=TestPlayerConnectWithCapacityZero
Step #1 - "e2e-feature-gates": time="2023-03-31 22:07:53.353" level=info msg="waiting for fleet condition" fleet=simple-fleet-1.02mm9l test=TestUpdateFleetReplicaAndSpec
Step #1 - "e2e-feature-gates": time="2023-03-31 22:07:53.410" level=info msg="Event!" lastTimestamp="2023-03-31 22:06:09 +0000 UTC" message="Port allocated" reason=PortAllocation test=TestPlayerConnectWithCapacityZero type=Normal
Step #1 - "e2e-feature-gates": time="2023-03-31 22:07:53.411" level=info msg="Event!" lastTimestamp="2023-03-31 22:06:09 +0000 UTC" message="Pod game-servers4l44 created" reason=Creating test=TestPlayerConnectWithCapacityZero type=Normal
Step #1 - "e2e-feature-gates": time="2023-03-31 22:07:53.411" level=info msg="Event!" lastTimestamp="2023-03-31 22:06:09 +0000 UTC" message="Address and port populated" reason=Scheduled test=TestPlayerConnectWithCapacityZero type=Normal
Step #1 - "e2e-feature-gates": time="2023-03-31 22:07:53.411" level=info msg="Event!" lastTimestamp="2023-03-31 22:07:42 +0000 UTC" message="SDK state change" reason=RequestReady test=TestPlayerConnectWithCapacityZero type=Normal
Step #1 - "e2e-feature-gates": time="2023-03-31 22:07:53.411" level=info msg="Event!" lastTimestamp="2023-03-31 22:07:42 +0000 UTC" message="SDK.Ready() complete" reason=Ready test=TestPlayerConnectWithCapacityZero type=Normal
```

Compared to my local repros, which also have:

```
time="2023-03-31 23:37:54.123" level=info msg="Event!" lastTimestamp="2023-03-31 23:36:34 +0000 UTC" message="Issue with Gameserver pod" reason=Unhealthy test=TestPlayerConnectWithCapacityZero type=Warning
```

but critically, on the e2e cluster there's a huge timing gap there with >90s between `Address and port populated` (so `Scheduled`) and `SDK state change`.

**Comment by zmerlynn** [comment]
Well.. it seems we are reusing the Pod, unintentionally. https://screenshot.googleplex.com/6pazQLj6uuHVj5n

**Comment by zmerlynn** [comment]
Looks like that cleared the hurdle - @markmandel if you could PTAL when you get a chance, thanks!

**Review by markmandel — APPROVED** [review]
Looks good to me! Nice work!

**Comment by markmandel** [comment]
Seeing lots of failures again 😞  @zmerlynn did you want to take a look?

**Comment by zmerlynn** [comment]
Yeah, I am looking. :(
